### DISCUSSION TITLE
Better code reuse for Meson with `Meson fragments` in Meson language.

### DISCUSSION THREAD

**Discussion by germandiagogomez** [mesonbuild/meson#13451] [Ideas]
Dear Meson team,

I have been working in my own projects in Meson for a while. FWIW, I would like to discuss an idea that came out from my own usage patterns over time about Meson for better code reuse and show an example of what I am doing currently in the context of a Conan Meson module that is not authored in Python.

I would like to draw a bit of your attention since I am aware of all public discussions (I believe) about adding includes, functions, etc. to Meson. I do not think this is a great idea since it could potentially obfuscate code, but, maybe by reusing much of what we have we can achieve something very close to abstraction and reuse in pure Meson,
 but in Meson directly.

I would like to discuss why I think enabling some kind of modules would be a good idea. Since the discussion can be potentially long and involved, I set this topic as a discussion.

# The code reuse problem

The real problem with Meson, which is also one of its strengths, is its abstraction level is relatively low but the language very readable: there are no functions, no macros and no way to reuse code easily in the sense of what CMake offers.

However, I am not here to propose any kind of functions or incudes, but to discuss what could be the best way forward to have a better level of code reuse.

For that, I would like to show what I am doing right now and discuss the trade-offs it has to use Python modules in my own opinion compared to authoring some kind of module in Meson.

## Meson strengths

I like that Meson does not have too much abstraction when writing files: that makes things easy to read. On the other hand, we cannot make everyone happy and people often demand functions or includes.

Functions, in some way, IMHO, can potentially obfuscate the code when they are not well-known and documented. Arbitrary-path includes break code structure and make path management difficult and project inspection more challenging. I think one of the strengths of Meson is that build files are always called `meson.build`, subprojects are always in `subprojects` directory and options are always in `meson.options` files. You go to a new project and no surprises, no tricks. 

This makes projects easy to navigate and code, even if a bit verbose at times, easy to follow and understand, together with the readable syntax that Meson has.

I would not change any of those but if I made some changes, it would be in the context of not obfuscating code and still mandating structure into how to organize project structure in general.

## Meson modules: current status

Meson currently has a set of modules, written in Python. Writing modules in Python is the way to go if you want new features into Meson.

Pragmatically speaking, I think that having to go to Python and the Meson repository just to add reusable code is a far cry for most developers that are just interested in a low-cost mechanism to reuse code without resorting to copy-paste and similar.

Most people are not going to author a Python module and go through all the process to make it into a Pull Request, get reviews, having the thumbs up from maintainers, etc. This is time-consuming and in many occasions not worth the effort.

However, the need for code reuse remains.

## Pure meson module (renamed to `Meson fragment`, better syntax pending for input/output vars) what I did and the patterns that emerged.

This is how I am currently implementing Meson "modules" on my own side. This is just a discussion on how to move things forward. I am not proposing anything beyond feedback at this point.

A Module, the way I am doing it, provides some `meson.options` to the user. Currently, in order to use those options, I paste them in the top-level directory.

Good things about having a "module" with Meson code:

     - I do not need to author anything in Python that needs to go through all the process into Meson project.
     - I can reuse my code (mostly, with `meson.options` caveats).

A Module has, in my own implementation:

         1. input variables. (optional and/or mandatory).
         2. output variables.
         3. regular meson options (which must be added to the meson.options file)
         
The good thing about modules is that they are regular Meson code, without new abstraction mechanisms. Code is clear to read once you open it, because Meson is easy to read.

A Module lives always in `$project_root/meson/modules/modulename` in my case, in `$project_root/meson/modules/conan`. The modulename directory contains a regular `meson.build` file.

## My Conan "module" (renamed to `Meson fragments` in the coming comments)

My conan module has all its documentation at the top, so please read and check, and follow after reading:

```
## This is  a meson conan  module authored  in pure meson  + a couple  of helper
## scripts.
##
## This  module calls  conan at  configuration time  via a  run_command command,
## invoking conan to build (conditionally)  and install packages information for
## consumption from your  own provided conanfile.txt or conanfile.py  at the top
## of your project.
##
## This module  does not  do any effort  in generating a  conanfile of  any kind
## whatsoever on your behalf.
##
## The  module detects  the correct  build type  conan needs  for your  packages
## leaning on your buildtype and debug option settings.
##
## Note that json generator in your  [generators] section in your conanfile must
## be used. in order for CONAN_BIN_PATHS  to work. Conan profiles should live in
## the root project directory in conan/profiles directory.
##
## These  are the  parameters  that are  available (add  these  options to  your
## meson.options file at the top-level so that they are forwarded to this module
## itself.
##
################################ INPUT PARAMETERS: #############################
##
## CONAN_PROFILES_MAPPING   (optional):  a   dictionary  that   uses  as   keys,
## host_machine.system()  names from  meson and  as  values a  filename list  in
## conan/profiles   directory.    The   first    profile   that    starts   with
## meson.get_compiler('cpp').get_id() will be used  for the mapping detection if
## 'auto' is used
##
##
#################### Options you need to add to the top-level of your project ##

# option('conan-remote', type: 'string', value: 'conancenter', description: 'the remote to download packages from')
# option('conan-profile', type : 'string', value: 'auto',
#        description: '''Name without path of conan profile to use. Available profiles
#                     are in conan/profiles. (default auto autodetects profile).''')

# option('build-machine-conan-profile', type : 'string',
#        value: 'same-as-host',
#        description: '''Use a profile for build machine. By default it is the same used for host. Must be set when cross-compiling.''')
# option('conan-build-packages', type : 'combo', choices: ['missing', 'never'],
#        value: 'never', description: 'Whether to build missing packages in Conan when they are missing or not.')
#
# option('conan-install-path', type : 'string',
#        value: 'auto', description: 'The install path for conan. Note that your project pkg_config_path must point to this path. Defaults to pkg_config_path if set')
#
############### OUTPUT PARAMETERS ##############################################
#
# CONAN_BIN_PATHS: binary  paths that conan  found when configuring  packages so
# that they can be used in find_program.  This is useful for tools that generate
# code in the build machine such as capnproto, protocol buffers and others.
#
#

# Note here how I use get_variable dynamically, because this input parameter is 
# optional
CONAN_PROFILES_MAPPING = get_variable('CONAN_PROFILES_MAPPING', {})
conan_prog = find_program('conan')

profiles_dir = meson.project_source_root() / 'conan/profiles/'
profilename = ''
if get_option('conan-profile') == 'auto'
  fs = import('fs')
  host_system = host_machine.system()
  cxx_compiler = meson.get_compiler('cpp')
  cxx_compiler_id = cxx_compiler.get_id()
  cxx_compiler_major_version = cxx_compiler.version().split('.')[0]

  maybe_profile = CONAN_PROFILES_MAPPING.get(host_system, '')
  if not maybe_profile.startswith(cxx_compiler_id)
    error('Could not autodetect a conan profile. Please check in @0@/conan/profiles/ if there is one that fits you or create a new one'.format(meson.project_source_root()) +
          ' and use it with -Dconan-profile=<your-profile>. You can also add it to CONAN_PROFILES_MAPPING for autodetection.')
  endif
  profilename = maybe_profile
else
  profilename = get_option('conan-profile')
endif


build_machine_profilename = profilename
if meson.is_cross_build()
  if get_option('build-machine-conan-profile') == 'same-as-host'
    available_profiles = run_command('scripts/listdir.py', meson.project_source_root() / 'conan/profiles', capture: true, check: true).stdout()
    error(f'You must set "build-machine-conan-profile" explicitly when cross-compiling. Available profiles:\n\n @available_profiles@')
  endif
else
  if get_option('build-machine-conan-profile') == 'same-as-host'
    build_machine_profilename = profilename
  else
    build_machine_profilename = get_option('build-machine-conan-profile')
  endif
endif

chosen_profile_path = meson.project_source_root() / f'conan/profiles/@profilename@'
chosen_build_machine_profile_path = meson.project_source_root() / f'conan/profiles/@build_machine_profilename@'

MESON_TO_CONAN_PROFILE = {'debug': 'Debug', 'release': 'Release', 'releasewithdebuginfo': 'Release',
                          'minsize': 'Release',
                          'plain': 'Release'}

# See https://blogs.gnome.org/mcatanzaro/2022/07/15/best-practices-for-build-options/
need_windows_debug_mode = host_machine.system() == 'windows' and get_option('debug')

CONAN_PROFILE_OPTIONS = {
   'profile': chosen_profile_path,
   'remote': get_option('conan-remote'),
   'profile:build': (meson.is_cross_build() ? chosen_build_machine_profile_path : chosen_profile_path),
   'build': get_option('conan-build-packages'),
   'buildtype': need_windows_debug_mode ? 'Debug' : MESON_TO_CONAN_PROFILE[get_option('buildtype')]
}

conan_profile_options_str = ''
foreach opt_key, opt_val : CONAN_PROFILE_OPTIONS
  conan_profile_options_str += f'@opt_key@=@opt_val@, '
endforeach
message(f'Using conan profile with options: @conan_profile_options_str@\n')

conan_install_path = ''
if get_option('pkg_config_path').length() != 0
  conan_install_path =  get_option('pkg_config_path')[0]
else
  error('conan-install-path option is empty. Set it or set pkg_config_path option and conan-install-path will use it.')
endif

_run_conan_cmd = [conan_prog.full_path(),
                  'install',
                  '--profile:host=@0@'.format(chosen_profile_path),
                  meson.project_build_root(),
                  '-r',
                  get_option('conan-remote'),
                  '-s:b',
                  'build_type=@0@'.format(CONAN_PROFILE_OPTIONS['buildtype']),
                  '-s',
                  'build_type=@0@'.format(CONAN_PROFILE_OPTIONS['buildtype']),
                  '-pr:b=' + (meson.is_cross_build() ? chosen_build_machine_profile_path : chosen_profile_path),
                  '--build=@0@'.format(get_option('conan-build-packages')),
                  '-if',
                  '@0@'.format(conan_install_path)]


conan_env = environment({'CONAN_SYSREQUIRES_MODE': 'verify'})
run_target('conan_install', command: _run_conan_cmd, env: conan_env)

conan_install_result = run_command(_run_conan_cmd, check: false, capture: true, env: conan_env)
if conan_install_result.returncode() != 0
  error('stdout\n---------------\n' + conan_install_result.stdout() + 'stderr\n---------------\n' + conan_install_result.stderr())
endif

pkg_cfg_path_conan_buildinfo = conan_install_path + '/conanbuildinfo.json'
CONAN_BIN_PATHS  = run_command('scripts/get_bin_dirs_from_buildinfo.py',
                                pkg_cfg_path_conan_buildinfo,
                                check: true).stdout().strip().split('\n')
```

## How I use the Meson "module"

At the top-level or any level of the project, I can use this module by doing:

```
# My optional input variable from the module

CONAN_PROFILES_MAPPING = {'darwin': 'clang15_libcxx', 'windows': 'msvc22', 'linux': 'gcc12_linux', 'emscripten': 'emscripten'}
subdir('meson/modules/conan')

# From here on, CONAN_BIN_PATHS is available, conan install installed its packages, etc.
## In another dir or wherever
capnp_compiler = find_program('capnpc', dirs: CONAN_BIN_PATHS, native: true)
capnp_cpp_plugin = find_program('capnpc-c++', dirs: CONAN_BIN_PATHS, native: true).full_path()
```

So basically, this has achieved code reuse, maybe with a couple of caveats:

       - `meson.options` still needs to be open to copy-paste project options from my own module
       - variables pollution?
       - inputs and outputs are documented, but it is just that, documentation. Not sure it is good or bad.

The module is basically data-driven, which is a good thing:

     1. You set your input variables.
     2. You set your options.
     3. You get your action/output variables.

In this sense, I think it is nicer than polluting your code with real user-defined functions around, though it could be seen as half-equivalent in some way. The usage from outside is intended to be through those two input mechanisms and and output variable mechanisms (probably even files actually...?), so the scope is well-defined.

## Module reuse and potential improvements

I can imagine a github repo (that I could even open myself!) and systematize these Meson modules without any new language features to be able to reuse code logic for this Conan module, installers and others.

This is not a proposal, just a discussion, though I think that, with minimal changes, we could have this mechanism standardized in some ways:

     1. A `meson.options` file could "include" the options from a module. Maybe by including a `module-<modulename>.meson.options` would introduce the options in the `meson.options` file?. Options in a module file must be at the top-level. Or maybe in a separate file but with a naming convention, in the Meson enforced structure style.
     2. From Meson code, be able to do `include('conan')` or whatever marks that as a module. Modules would always live in `meson/modules` directory or a fixed directory, to not break layout and to make it discoverable. Note that modules do not introduce any kind of subproject mechanism in any way.
     3. Including a module is the same as including a regular meson file except for the fact that it can have options and they must be the first thing to be introduced (if they go in the same file) or split apart and with a name convention.

Maybe input variables could be marked in some way? Not sure it is even needed. It works as it is currently but they are not marked in any way.

This "module" method already works, just by making it minimally prettier and without messing with Meson rules or very elaborate stuff, it is perfectly usable IMHO.

I think this has the advantage that modules can be collected and reused for projects more easily. It also has the advantage that it uses well-known features and mechanisms yet it does not add new rules or anything similar to the language itself, just a couple of ways to get the options and better encapsulation.

What do you think? Problems? Improvements?

**Comment by germandiagogomez** [comment]
First problem I found: how to invoke the module several times.

I will keep working in ideas to this as things evolve. Probably a module should be something unrelated to meson options (not sure yet) and should be some kind of meson code like a black box that communicates through input and output variables only, in a way that it does not pollute the environment around and that it can be invoked many times. Something along these lines.

In this model modules can be used many times and communicate through variables (still flowing everything as an idea, feeedback is very welcome). This is similar to a function, I would say, but without introducing functions themselves. Though it can also add targets, which functions cannot.

In this model options are independent of the module, variables are isolated in their own module file:
```
conan_invocation = use_module('conan', input: [{'CONAN_PROFILES_MAPPING': {'darwin': 'clang15_libcxx', 'windows': 'msvc22', 'linux': 'gcc12_linux', 'emscripten': 'emscripten'}, {'CONAN_REMOTE': get_option('conan-remote')}, ...])

CONAN_BIN_PATHS = conan_invocation.get_output_variable('CONAN_BIN_PATHS')
```

In this model modules only have input and output and do not pollute the outer world, though they can have side-effects such as configuring a file or similar.

**Comment by tristan957** [comment]
@dcbaker what was the functools.partial type thing you were working on?

**Reply by dcbaker** [reply]
My thought was having functions where you could pre-set some of the arguments and fill them in later, something like:
```meson
templ = template(
  custom_target,
  command : [find_program('generator'), '@INPUT1@', '@OUTPUT0@', '--argument-file', '@INPUT0@']
  install : true,
  install_tag : 'runtime',
  build_by_default : true,
)

x = templ.call(
  input : ['my.file', 'other.file'],
  output : ['output.file'],
)

y = templ.call(
  input : ['file.2', 'file.3'],
  output : ['out.file']
)
```

**Reply by dcbaker** [reply]
My design is really about argument re-use

**Comment by germandiagogomez** [comment]
## More ideas

So, taking a look at the initial solution, some evolutionary comments here I have been thinking about:

I think an appropriate name for this feature could be 'Meson fragments'.

What is a fragment? A fragment is a reusable piece of Meson code in which some input variables are replaced by variables provided by the user and the code is invoked.

`Meson fragments` are templates for the code to be run, but not templates with different syntax: just regular Meson code. `Fragment results` are the result of invoking a fragment.

  - A fragment is a black box. No variable that is explicitly marked as input or output is seen outside of the scope of the fragment. Input and output variables must be marked at the top-level of the Meson fragment.
  
  - A fragment has input variables which can be of any kind that Meson supports.
  - A fragment has output variables and potential side-effects.
  - A fragment can be reused many times with different parameters and be called repeteadly. The call to a fragment is a `fragment result`.
  - Each fragment lives in a fragment file, always, and in a fixed directoy, a-la `subprojects`, for easy inspection.
  - Fragments do not call other fragments or subdir. But can call python scripts or any other function that Meson allows in its API.
  - A fragment can add targets.

The goal of a fragment is to perform actions that currently take quite a bit of boilerplate and cannot be reused except by copy-paste.

## What are the differences between a fragment and regular functions?

A fragment, as functions, has input and output variables and are black boxes. The input variables of a fragment and the output parameters of a fragment can be marked in the Meson fragment.

Code reuse can be achieved with regular Meson code. The top-level code is familiar and uses well-known functions from the Meson API. 

Unlike functions, code fragments can add targets. 

## Use cases in my own code

### Creating an installer (fragment consumer side syntax)

```
installer_fragment_result = use_fragment('create_nstaller', {'Targets': [my_exe, mylib],...)

installer_tgt = installer_fragment_result.get_output('installer')

# Can use result to upload artifact, for example.
```


### Create a Meson fragment to use Conan

Same idea here. The exposed code in my first comment could use conan, input vars and add a run target like this: 

```
### Config-time conan invocation, run target for invoking conan added.
### Not clear if the run target should be added to
the top-level targets though.
conan_frag_result = use_fragment('conan', input: {'conan_profile': ...})

```

## Potential syntax for fragment creation 

In the next days I will post how the syntax for code fragment creation looks. It is basically the same with some replacement variables marked as inout in some way and with output vars also set.

There needs to be a way to mark variables as inout and output. Probably with a new API that declares variables as fragment input vars and output vars and the rest of the code identical to regular Meson code it could be enough.

Need to give some thought  about it.

**Comment by annacrombie** [comment]
[muon](https://github.com/annacrombie/muon) has support for writing modules in a slightly extended meson syntax: https://github.com/annacrombie/muon/blob/master/doc/script_modules.md

Here is the i18n module implemented this way: https://github.com/annacrombie/muon/blob/master/src/script/modules/i18n.meson.

Currently these modules must be built-in to muon so must go through the normal review process for any code, but it would be simple to relax that constraint.

**Reply by germandiagogomez** [reply]
@annacrombie thanks for the links I was not aware of those.

There has been a lot of heated debate against adding functions in Meson language, including @jpakkane explicit rejection as far as my knowledge goes. I see Meson functions in Muon so I suspect that it would be rejected.

That is why I am proposing fragments. Fragments are meant to be isolated pieces that have some differences with functions and are meant to be used coarse-grained. I see fragments as "enclosed normal meson.build syntax" into its own file that can be reused as a function.

If instead we have functions, I think there would be a bigger tendency to write special-case functions for each small use of things. That would lead to a proloferation of APIs. Being fragments heavier and more coarse-grained, they are supposed to do something bigger but when you open a fragment you just see normal Meson functions, including the possibility of adding targets.

I think this would play in favor of keeping things readable. My experience in CMake is that it becomes programming. In Meson my experience is that things are very uniform (and that is excellent for readability) so the goal in my mind is to enable code resuse but respecting the original APIs and making use of those. If we add functions, the result will be new APIs on top.

I still think that for new full APIs a Python module such as qt module or what already exists is the better way. Fragments would be like "pack these 100 lines of code and make them reusable", but those 100 lines of code will not call other fragments. Things stay almost "flat", which contributes to readability.

Not even 100% sure if what I am discussing here since it is evolving, but I like the idea of adding minimal syntax, keeping fragments in a single place, keeping the code inside fragments flat and top level as regular Meson syntax and restricting invocation (fragments do not call other fragments) which would lead to a call graph and goes against readability.

**Reply by eli-schwartz** [reply]
Probably the biggest issue to my mind with functions as a generic concept is, indeed, that implementing them in a straightforward way implies adding infinite recursion to the language. I would say that regardless of anything else, the meson maintainers will *never* change their minds on this.

I would not rule out the possibility of meson one day having some form of non-recursive logic reuse. :) Not yet sure what it would look like of course!

**Reply by germandiagogomez** [reply]
@eli-schwartz I agree with Meson maintainers about the recursion stuff.  I would not add recursion or arbitraty path loading of things and such. Things should be restricted and located always in the same places. Those make things far easier in the long run.

One of the good things from Meson is that we all read the same code such as custom target, executable, test, etc. No tricks. We are all familiar with it. With regular functions things would change that fact: recursion makes reading harder and arbitraty path loading makes discoverability worse.

We also find projects in subprojects dir. We find options in meson.options file.

You go to a foreigner project and things are obvious. I cannot say the same about other build systems.

That is the essence I would like to keep with something like fragments. I would outlaw functions directly in fact: they tend to be more fine-grained, creating new APIs proliferating in all files. They also add new syntax and rules.

I think it is simpler to have a piece of current Meson code an say: these are the inputs and this is how to invoke it.

Maybe a dcitionary of input and output variables at the top and regular Meson code with a directory for fragments and a single `invocation = use_fragment('myfragment', inputs: {})` is enough. 

No special cases, no new functions, just Meson code. Nothing. Just that and a directory. And variables shielded from the outer world except for declared input and output variables.

For regular functions design, I would still stick to Python modules and on purpose: this makes the barrier to sdd new fine-grained APIs more difficult, discouraging proliferation of unreadable code by others.

**Reply by annacrombie** [reply]
> implementing them in a straightforward way implies adding infinite recursion to the language

It is trivial to disallow this if desired.  On every non-native function invocation simply walk the call stack and check if that function has been called before.  This is exactly how meson currently disallows infinite recursion via the `subdir()` call currently (`subdir()` can be thought of as a very primitive function call with no scope).

> Maybe a dcitionary of input and output variables at the top and regular Meson code with a directory for fragments and a single invocation = use_fragment('myfragment', inputs: {}) is enough.

This is actually just a crippled way to emulate arguments and return values to a function where the function is the entire meson source file.

**Reply by germandiagogomez** [reply]
> This is actually just a crippled way to emulate arguments and return values to a function where the function is the entire meson source file.

It could look to you like that, but it has other advantages. For example, it does not add function prototypes and syntax + new rules to the language for that. It also encourages coarse-grained functionality as opposed to the proliferation of new "APIs" in Meson files. I would think of this last thing as a feature, since functions are easily abused for creating new APIs at the slightest chance.

So the way I would see it would be: want new functions? Add to a Python module or Meson directly (with its process review, etc) and make them well-known to everyone.

Want to avoid boilerplate for your own project(s) but with still readable code for projects and to some extent reusable? Use fragments.

The fact that fragments go in a separate file add to the effort to add them to a project, promoting using existing (and understandable) APIs inside those at the top-level. They go in a feagments subdir. Of course they could also be abused as functions in disguise but that design signals them as something heavier.

Add random functions and you will end up with DSLs all around instead of regular Meson well-known calls getting nearly equivalent things to what I saw in many CMake projects (minus the recursion).

It is easy to creat 10 or 15 functions in a small project with functions that you do not know as a foreign user. But you should not see small projects with 10 or 15 fragments if they are this "heavy" to create, it is just less heavy to call well-known APIs inline.

I see that as a feature taking into account that targetting reading uniformity and project structure is a good thing of any Meson project.

If you want really involved, fine-grained functions, add them to Meson and promote them as well-known APIs. Otherwise you end up wondering what every call does at every step of reading a Meson file and that goes against Meson readability altogether. At the slightest thing it should be known where a function or fragment is invoked from, clearly. 

I also suspect that since functions have scope, return types and other stuff, they must be more carefully designed by thinking of new rules. Fragments are just regular code in another file shielded. No function prototypes, not rules for what they return or not. Nothing. Just "use these variables as input" and "use these variables as output". Also, it seems more natural to define targets in those fragments than inside functions.

Just my two cents.

**Reply by ferdnyc** [reply]
(In order:) @germandiagogomez, @annacrombie, @germandiagogomez 

> > > Maybe a dcitionary of input and output variables at the top and regular Meson code with a directory for fragments and a single invocation = use_fragment('myfragment', inputs: {}) is enough.
> > 
> > This is actually just a crippled way to emulate arguments and return values to a function where the function is the entire meson source file.
> 
> It could look to you like that, but it has other advantages. For example, it does not add function prototypes and syntax + new rules to the language for that. It also encourages coarse-grained functionality as opposed to the proliferation of new "APIs" in Meson files. I would think of this last thing as a feature, since functions are easily abused for creating new APIs at the slightest chance.

I don't know, I kind of agree with @annacrombie here.

Defining a fragment as a "file" which takes "inputs" (whether via existing global variables, or arguments passed to the import call... or both) and produces "outputs" (in the form of global symbols created by the fragment, I presume?) just makes each fragment file a function that operates _entirely_ through side effects, which is... well, it's one approach, but generally the idea is to **minimize** side effects from calling other code, not make them the central aspect of a language feature.

And if the fragment file implements the entire Meson language, what's to stop a fragment file from containing `use_fragment()` calls of its own? Then you've got the outer fragment producing _undocumented_ side effects, unless it incorporates the outputs documentation for all of the fragments _it_ calls as well. (Plus, again, the potential for infinite recursion.)

From a scoping perspective and to avoid collisions in the global variable namespace, I'd much prefer your suggestion of assigning the output of the fragment to a variable in the calling scope — so it acts kind of like a Meson module with no methods, just static data. And even better if that's the _only_ way for it to output data, and any variables set during its invocation are discarded (or automatically redirected to the object returned by the invocation).

The question of what exactly should happen for other side-effect actions inside fragments, like if it creates one or more build targets, is still a tricky one.

And pretending these fragments aren't now functions feels like just that: pretending.

(You can technically build the equivalent of fragments in CMake, since every `include()`d file will be immediately executed as a CMake script. They don't _need_ to contain or use functions. The reason they often **DO** contain function definitions, and sometimes exclusively contain function definitions, is to minimize the side effects of the `include()`. If an `include()`d script only defines functions, then it leaves the _invocation_ of that code up to the caller, instead of immediately triggering all the code as a side effect of the `include()` call. That's generally considered a positive, not something to be avoided.)
